In [1]:
from google.colab import drive

In [2]:
drive.mount('/content/drive')

Mounted at /content/drive


#  **ETL for functions 3, 4, and 5**

Now, with the datasets unnested and organized, we are going to do some merges and data trasnformations in order to have the information required in a CSV file to develop these functions and models:


3. **UsersRecommend**(`year`: int) Function:

  It returns the top 3 games MOST recommended by users for the given `year` (reviews.recommend = True and positive/neutral comments).

  Example of return: [{"Rank 1" : X}, {"Rank 2" : Y},{"Rank 3" : Z}]

4. **UsersNotRecommend**(`year`: int) Function:

  It returns the top 3 games LEAST recommended by users for the given `year` (reviews.recommend = False and negative comments).

  Example of return: [{"Rank 1" : X}, {"Rank 2" : Y},{"Rank 3" : Z}]

5. **Sentiment_analysis**(`year`: int) Function:

  Based on the release `year`, it returns a list with the count of user review records categorized with sentiment analysis.

  Example of return: {"Negative": 182, "Neutral": 120, "Positive": 278}
  
6. **Recommendation System**

- If it's an item-item recommendation system:

  def recommendation_game(`product id`):
  By entering the product id, we should receive a list of 5 recommended games similar to the one entered.

- If it's a user-item recommendation system:

  def recommendation_user(`user id`):
  By entering a user id, we should receive a list of 5 recommended games for that user.

### 1. Data Load:

In [3]:
import pandas as pd
import numpy as np

In [4]:
df_steam_games = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/steam_games_final.csv')
df_steam_games

publisher                                             genres  \
0             Kotoshiro  ['Action', 'Casual', 'Indie', 'Simulation', 'S...   
1      Making Fun, Inc.       ['Free to Play', 'Indie', 'RPG', 'Strategy']   
2          Poolians.com  ['Casual', 'Free to Play', 'Indie', 'Simulatio...   
3                  彼岸领域                  ['Action', 'Adventure', 'Casual']   
4                   NaN                                                NaN   
...                 ...                                                ...   
32130   Ghost_RUS Games      ['Casual', 'Indie', 'Simulation', 'Strategy']   
32131            Sacada                    ['Casual', 'Indie', 'Strategy']   
32132      Laush Studio                  ['Indie', 'Racing', 'Simulation']   
32133          SIXNAILS                                ['Casual', 'Indie']   
32134               NaN                                                NaN   

                       app_name                     title  \
0           Lost Summoner Kitty       Lost Summoner Kitty   
1                     Ironbound                 Ironbound   
2       Real Pool 3D - Poolians   Real Pool 3D - Poolians   
3                       弹炸人2222                   弹炸人2222   
4                 Log Challenge                       NaN   
...                         ...                       ...   
32130            Colony On Mars            Colony On Mars   
32131  LOGistICAL: South Africa  LOGistICAL: South Africa   
32132             Russian Roads             Russian Roads   
32133       EXIT 2 - Directions       EXIT 2 - Directions   
32134               Maze Run VR                       NaN   

                                                     url release_date  \
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
1      http://store.steampowered.com/app/643980/Ironb...   2018-01-04   
2      http://store.steampowered.com/app/670290/Real_...   2017-07-24   
3         http://store.steampowered.com/app/767400/2222/   2017-12-07   
4      http://store.steampowered.com/app/773570/Log_C...          NaN   
...                                                  ...          ...   
32130  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
32131  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
32132  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
32133  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
32134  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                    tags  \
0      ['Strategy', 'Action', 'Indie', 'Casual', 'Sim...   
1      ['Free to Play', 'Strategy', 'Indie', 'RPG', '...   
2      ['Free to Play', 'Simulation', 'Sports', 'Casu...   
3                      ['Action', 'Adventure', 'Casual']   
4                ['Action', 'Indie', 'Casual', 'Sports']   
...                                                  ...   
32130      ['Strategy', 'Indie', 'Casual', 'Simulation']   
32131                    ['Strategy', 'Indie', 'Casual']   
32132                  ['Indie', 'Simulation', 'Racing']   
32133  ['Indie', 'Casual', 'Puzzle', 'Singleplayer', ...   
32134  ['Early Access', 'Adventure', 'Indie', 'Action...   

                                             reviews_url  \
0      http://steamcommunity.com/app/761140/reviews/?...   
1      http://steamcommunity.com/app/643980/reviews/?...   
2      http://steamcommunity.com/app/670290/reviews/?...   
3      http://steamcommunity.com/app/767400/reviews/?...   
4      http://steamcommunity.com/app/773570/reviews/?...   
...                                                  ...   
32130  http://steamcommunity.com/app/773640/reviews/?...   
32131  http://steamcommunity.com/app/733530/reviews/?...   
32132  http://steamcommunity.com/app/610660/reviews/?...   
32133  http://steamcommunity.com/app/658870/reviews/?...   
32134  http://steamcommunity.com/app/681550/reviews/?...   

                         

In [5]:
df_steam_games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32135 entries, 0 to 32134
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   publisher     24073 non-null  object 
 1   genres        28852 non-null  object 
 2   app_name      32133 non-null  object 
 3   title         30085 non-null  object 
 4   url           32135 non-null  object 
 5   release_date  30068 non-null  object 
 6   tags          31972 non-null  object 
 7   reviews_url   32133 non-null  object 
 8   specs         31465 non-null  object 
 9   price         30758 non-null  object 
 10  early_access  32135 non-null  bool   
 11  id            32133 non-null  float64
 12  developer     28836 non-null  object 
dtypes: bool(1), float64(1), object(11)
memory usage: 3.0+ MB


In [6]:
df_user_reviews = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/user_reviews_final.csv')
df_user_reviews

user_id                                           user_url  \
0      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
2      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
3                js41637               http://steamcommunity.com/id/js41637   
4                js41637               http://steamcommunity.com/id/js41637   
...                  ...                                                ...   
59328  76561198312638244  http://steamcommunity.com/profiles/76561198312...   
59329  76561198312638244  http://steamcommunity.com/profiles/76561198312...   
59330        LydiaMorley           http://steamcommunity.com/id/LydiaMorley   
59331        LydiaMorley           http://steamcommunity.com/id/LydiaMorley   
59332        LydiaMorley           http://steamcommunity.com/id/LydiaMorley   

                                  funny                     posted  \
0                                   NaN   Posted November 5, 2011.   
1                                   NaN      Posted July 15, 2011.   
2                                   NaN     Posted April 21, 2011.   
3                                   NaN      Posted June 24, 2014.   
4                                   NaN  Posted September 8, 2013.   
...                                 ...                        ...   
59328                               NaN            Posted July 10.   
59329                               NaN             Posted July 8.   
59330  1 person found this review funny             Posted July 3.   
59331                               NaN            Posted July 20.   
59332                               NaN             Posted July 2.   

      last_edited   item_id                                          helpful  \
0             NaN    1250.0                                   No ratings yet   
1             NaN   22200.0                                   No ratings yet   
2             NaN   43110.0                                   No ratings yet   
3             NaN  251610.0  15 of 20 people (75%) found this review helpful   
4             NaN  227300.0     0 of 1 people (0%) found this review helpful   
...           ...       ...                                              ...   
59328         NaN      70.0                                   No ratings yet   
59329         NaN  362890.0                                   No ratings yet   
59330         NaN  273110.0    1 of 2 people (50%) found this review helpful   
59331         NaN     730.0                                   No ratings yet   
59332         NaN     440.0                                   No ratings yet   

      recommend  sentiment_analysis  
0          True                   2  
1          True                   2  
2          True                   2  
3          True                   2  
4          True                   2  
...         ...                 ...  
59328      True                   2  
59329      True                   2  
59330      True                   2  
59331      True                   2  
59332      True                   2  

[59333 rows x 9 columns]

In [7]:
df_user_reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59333 entries, 0 to 59332
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   user_id             59333 non-null  object 
 1   user_url            59333 non-null  object 
 2   funny               8151 non-null   object 
 3   posted              59305 non-null  object 
 4   last_edited         6140 non-null   object 
 5   item_id             59305 non-null  float64
 6   helpful             59305 non-null  object 
 7   recommend           59305 non-null  object 
 8   sentiment_analysis  59333 non-null  int64  
dtypes: float64(1), int64(1), object(7)
memory usage: 4.1+ MB


In [8]:
df_users_items = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/users_items_final.csv')
df_users_items

user_id  items_count           steam_id  \
0        76561197970982479          277  76561197970982479   
1        76561197970982479          277  76561197970982479   
2        76561197970982479          277  76561197970982479   
3        76561197970982479          277  76561197970982479   
4        76561197970982479          277  76561197970982479   
...                    ...          ...                ...   
5170010  76561198329548331            7  76561198329548331   
5170011  76561198329548331            7  76561198329548331   
5170012  76561198329548331            7  76561198329548331   
5170013  76561198329548331            7  76561198329548331   
5170014  edward_tremethick            0  76561198331598578   

                                                  user_url   item_id  \
0        http://steamcommunity.com/profiles/76561197970...      10.0   
1        http://steamcommunity.com/profiles/76561197970...      20.0   
2        http://steamcommunity.com/profiles/76561197970...      30.0   
3        http://steamcommunity.com/profiles/76561197970...      40.0   
4        http://steamcommunity.com/profiles/76561197970...      50.0   
...                                                    ...       ...   
5170010  http://steamcommunity.com/profiles/76561198329...  373330.0   
5170011  http://steamcommunity.com/profiles/76561198329...  388490.0   
5170012  http://steamcommunity.com/profiles/76561198329...  521570.0   
5170013  http://steamcommunity.com/profiles/76561198329...  519140.0   
5170014     http://steamcommunity.com/id/edward_tremethick       NaN   

                             item_name  playtime_forever  playtime_2weeks  
0                       Counter-Strike               6.0              0.0  
1                Team Fortress Classic               0.0              0.0  
2                        Day of Defeat               7.0              0.0  
3                   Deathmatch Classic               0.0              0.0  
4            Half-Life: Opposing Force               0.0              0.0  
...                                ...               ...              ...  
5170010                    All Is Dust               0.0              0.0  
5170011  One Way To Die: Steam Edition               3.0              3.0  
5170012          You Have 10 Seconds 2               4.0              4.0  
5170013                     Minds Eyes               3.0              3.0  
5170014                            NaN               NaN              NaN  

[5170015 rows x 8 columns]

In [9]:
df_users_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5170015 entries, 0 to 5170014
Data columns (total 8 columns):
 #   Column            Dtype  
---  ------            -----  
 0   user_id           object 
 1   items_count       int64  
 2   steam_id          int64  
 3   user_url          object 
 4   item_id           float64
 5   item_name         object 
 6   playtime_forever  float64
 7   playtime_2weeks   float64
dtypes: float64(3), int64(2), object(3)
memory usage: 315.6+ MB


**Step 1:**

To find the years when the items were played we need to check the column `posted` of the dataframe df_user_reviews. We need to organize that dataframe in order to extract the year. Also, in this dataframe we have the column `item_id` that can help us to relate this information with the other dataframes.

First, let's review if there are some entries with no reviews. A way to do it is to look for NaN in the `posted` column.

In [10]:
mask1 = df_user_reviews['posted'].isna()
df_user_reviews[mask1]

user_id                                           user_url  \
137                gdxsd                 http://steamcommunity.com/id/gdxsd   
177    76561198094224872  http://steamcommunity.com/profiles/76561198094...   
2559   76561198021575394  http://steamcommunity.com/profiles/76561198021...   
10080            cmuir37               http://steamcommunity.com/id/cmuir37   
13767          Jaysteeny             http://steamcommunity.com/id/Jaysteeny   
15493             ML8989                http://steamcommunity.com/id/ML8989   
19184  76561198079215291  http://steamcommunity.com/profiles/76561198079...   
20223  76561198079342142  http://steamcommunity.com/profiles/76561198079...   
25056  76561198061996985  http://steamcommunity.com/profiles/76561198061...   
26257  76561198108286351  http://steamcommunity.com/profiles/76561198108...   
26674     CallTripleZero        http://steamcommunity.com/id/CallTripleZero   
28444       Priceless612          http://steamcommunity.com/id/Priceless612   
28452      diabolical666         http://steamcommunity.com/id/diabolical666   
28975         DrScottnik            http://steamcommunity.com/id/DrScottnik   
31955  76561198110176420  http://steamcommunity.com/profiles/76561198110...   
32819         jojoPL1987            http://steamcommunity.com/id/jojoPL1987   
33785        Kylee_Kylie           http://steamcommunity.com/id/Kylee_Kylie   
39279             _samie                http://steamcommunity.com/id/_samie   
39357   FifthExclamation      http://steamcommunity.com/id/FifthExclamation   
39420             sbenji                http://steamcommunity.com/id/sbenji   
40661     SpatialWheel14        http://steamcommunity.com/id/SpatialWheel14   
41777  76561198065238937  http://steamcommunity.com/profiles/76561198065...   
44384  76561198071075080  http://steamcommunity.com/profiles/76561198071...   
50718  76561198063489150  http://steamcommunity.com/profiles/76561198063...   
53323   thecommunistbear      http://steamcommunity.com/id/thecommunistbear   
55425  76561198085883839  http://steamcommunity.com/profiles/76561198085...   
57578  76561198100396147  http://steamcommunity.com/profiles/76561198100...   
57766  76561198102423654  http://steamcommunity.com/profiles/76561198102...   

      funny posted last_edited  item_id helpful recommend  sentiment_analysis  
137     NaN    NaN         NaN      NaN     NaN       NaN                   1  
177     NaN    NaN         NaN      NaN     NaN       NaN                   1  
2559    NaN    NaN         NaN      NaN     NaN       NaN                   1  
10080   NaN    NaN         NaN      NaN     NaN       NaN                   1  
13767   NaN    NaN         NaN      NaN     NaN       NaN                   1  
15493   NaN    NaN         NaN      NaN     NaN       NaN                   1  
19184   NaN    NaN         NaN      NaN     NaN       NaN                   1  
20223   NaN    NaN         NaN      NaN     NaN       NaN                   1  
25056   NaN    NaN         NaN      NaN     NaN       NaN                   1  
26257   NaN    NaN         NaN      NaN     NaN       NaN                   1  
26674   NaN    NaN         NaN      NaN     NaN       NaN                   1  
28444   NaN    NaN         NaN      NaN     NaN       NaN                   1  
28452   NaN    NaN         NaN      NaN     NaN       NaN                   1  
28975   NaN    NaN         NaN      NaN     NaN       NaN                   1  
31955   NaN    NaN         NaN      NaN     NaN       NaN                   1  
32819   NaN    NaN         NaN      NaN     NaN       NaN                   1  
33785   NaN    NaN         NaN      NaN     NaN       NaN                   1  
39279   NaN    NaN         NaN      NaN     NaN       NaN                   1  
39357   NaN    NaN         NaN      NaN     NaN       NaN                   1  
39420   NaN    NaN         NaN      NaN     NaN       NaN                   1  
40661   NaN    NaN         NaN      NaN     N

In [11]:
np.shape(df_user_reviews[mask1])

(28, 9)

We can see that there are 28 entries that does not have a review. We can eliminate those entries from our dataframe in order to have less noise in our analysis.

In [12]:
mask2 = ~df_user_reviews['posted'].isna()
df_user_reviews_NoNaN = df_user_reviews[mask2]
df_user_reviews_NoNaN.reset_index(drop=True, inplace=True)
df_user_reviews_NoNaN


user_id                                           user_url  \
0      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
2      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
3                js41637               http://steamcommunity.com/id/js41637   
4                js41637               http://steamcommunity.com/id/js41637   
...                  ...                                                ...   
59300  76561198312638244  http://steamcommunity.com/profiles/76561198312...   
59301  76561198312638244  http://steamcommunity.com/profiles/76561198312...   
59302        LydiaMorley           http://steamcommunity.com/id/LydiaMorley   
59303        LydiaMorley           http://steamcommunity.com/id/LydiaMorley   
59304        LydiaMorley           http://steamcommunity.com/id/LydiaMorley   

                                  funny                     posted  \
0                                   NaN   Posted November 5, 2011.   
1                                   NaN      Posted July 15, 2011.   
2                                   NaN     Posted April 21, 2011.   
3                                   NaN      Posted June 24, 2014.   
4                                   NaN  Posted September 8, 2013.   
...                                 ...                        ...   
59300                               NaN            Posted July 10.   
59301                               NaN             Posted July 8.   
59302  1 person found this review funny             Posted July 3.   
59303                               NaN            Posted July 20.   
59304                               NaN             Posted July 2.   

      last_edited   item_id                                          helpful  \
0             NaN    1250.0                                   No ratings yet   
1             NaN   22200.0                                   No ratings yet   
2             NaN   43110.0                                   No ratings yet   
3             NaN  251610.0  15 of 20 people (75%) found this review helpful   
4             NaN  227300.0     0 of 1 people (0%) found this review helpful   
...           ...       ...                                              ...   
59300         NaN      70.0                                   No ratings yet   
59301         NaN  362890.0                                   No ratings yet   
59302         NaN  273110.0    1 of 2 people (50%) found this review helpful   
59303         NaN     730.0                                   No ratings yet   
59304         NaN     440.0                                   No ratings yet   

      recommend  sentiment_analysis  
0          True                   2  
1          True                   2  
2          True                   2  
3          True                   2  
4          True                   2  
...         ...                 ...  
59300      True                   2  
59301      True                   2  
59302      True                   2  
59303      True                   2  
59304      True                   2  

[59305 rows x 9 columns]

Now, we need to extract the year of the column 'posted' in the dataframe df_user_reviews_NoNaN. But first, let's check the way the infomation is oragnized.

In [13]:
df_user_reviews_NoNaN['posted'][:60]


0       Posted November 5, 2011.
1          Posted July 15, 2011.
2         Posted April 21, 2011.
3          Posted June 24, 2014.
4      Posted September 8, 2013.
5      Posted November 29, 2013.
6             Posted February 3.
7       Posted December 4, 2015.
8       Posted November 3, 2014.
9       Posted October 15, 2014.
10      Posted October 15, 2014.
11      Posted October 15, 2014.
12      Posted October 14, 2013.
13         Posted July 28, 2012.
14          Posted June 2, 2012.
15         Posted June 29, 2014.
16     Posted November 22, 2012.
17     Posted February 23, 2012.
18        Posted April 15, 2014.
19     Posted December 23, 2013.
20        Posted March 14, 2014.
21         Posted July 11, 2013.
22           Posted May 5, 2014.
23     Posted December 24, 2012.
24      Posted October 21, 2012.
25        Posted March 20, 2012.
26         Posted March 9, 2012.
27                Posted May 20.
28               Posted July 24.
29      Posted February 1, 2015.
30        

We can see that there is a pattern in the organization of the data. Let's use the library re. The 're' library in Python is used for working with regular expressions. Regular expressions are powerful tools for pattern matching and text manipulation. Let's see an example:

In [14]:
import re

# Original list example:
original_list = ['Posted November 5, 2011.', 'Posted July 15, 2011.', 'Posted June 16', 'Posted March 30, 2015']

# List to store the years
years_list = []

# Regular expression to search for the year in 'YYYY' format
year_pattern = r'\b\d{4}\b'

for item in original_list:
    # Search for the year in each element of the list
    match = re.search(year_pattern, item)

    if match:
        # If a year is found, add it to the list of years
        years_list.append(match.group())
    else:
        # If no year is found, add 'NaN' to the list of years
        years_list.append('NaN')

# Print the list of years
print(years_list)



['2011', '2011', 'NaN', '2015']


Now, let's do it with the complete dataframe:

In [15]:
import re

# Original list:
original_list = df_user_reviews_NoNaN['posted']

# List to store the years
years_list = []

# Regular expression to search for the year in 'YYYY' format
year_pattern = r'\b\d{4}\b'

for item in original_list:
        # If the item is a string, search for the year
        match = re.search(year_pattern, item)
        if match:
            # If a year is found, add it to the list of years
            years_list.append(match.group())
        else:
            # If no year is found, add 'NaN' to the list of years
            years_list.append('NaN')

# Print the list of years
years_list[:30]


['2011',
 '2011',
 '2011',
 '2014',
 '2013',
 '2013',
 'NaN',
 '2015',
 '2014',
 '2014',
 '2014',
 '2014',
 '2013',
 '2012',
 '2012',
 '2014',
 '2012',
 '2012',
 '2014',
 '2013',
 '2014',
 '2013',
 '2014',
 '2012',
 '2012',
 '2012',
 '2012',
 'NaN',
 'NaN',
 '2015']

We can see that there are some NaN that do not have a year. It may happen because those reviews were made during the current year when the data was extracted. Let's check the maximuum year of the list obtained above.

In [16]:
# First, we need to filters the 'NaN' and convert the years to integers
filtered_years = [int(year) for year in years_list if year != 'NaN']

# Find the maximum year in teh filtered list
max_year = max(filtered_years)

# Print the maximum year
print(f"Maximum Year: {max_year}")

Maximum Year: 2015


We can check that the maximum year is 2015, so the year of the NaN posted dates is 2016. Let's change NaN values by 2016.

In [17]:
updated_year_list = ['2016' if year == 'NaN' else year for year in years_list]
updated_year_list[:30]

['2011',
 '2011',
 '2011',
 '2014',
 '2013',
 '2013',
 '2016',
 '2015',
 '2014',
 '2014',
 '2014',
 '2014',
 '2013',
 '2012',
 '2012',
 '2014',
 '2012',
 '2012',
 '2014',
 '2013',
 '2014',
 '2013',
 '2014',
 '2012',
 '2012',
 '2012',
 '2012',
 '2016',
 '2016',
 '2015']

In [18]:
np.shape(updated_year_list)

(59305,)

In [19]:
np.shape(df_user_reviews_NoNaN)

(59305, 9)

Finally, let's concatenate this list to our dataframe df_user_reviews_NoNAN:

In [20]:
df_user_reviews_WY = pd.concat([df_user_reviews_NoNaN, pd.DataFrame({'review_year': updated_year_list})], axis=1)
df_user_reviews_WY

user_id                                           user_url  \
0      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
2      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
3                js41637               http://steamcommunity.com/id/js41637   
4                js41637               http://steamcommunity.com/id/js41637   
...                  ...                                                ...   
59300  76561198312638244  http://steamcommunity.com/profiles/76561198312...   
59301  76561198312638244  http://steamcommunity.com/profiles/76561198312...   
59302        LydiaMorley           http://steamcommunity.com/id/LydiaMorley   
59303        LydiaMorley           http://steamcommunity.com/id/LydiaMorley   
59304        LydiaMorley           http://steamcommunity.com/id/LydiaMorley   

                                  funny                     posted  \
0                                   NaN   Posted November 5, 2011.   
1                                   NaN      Posted July 15, 2011.   
2                                   NaN     Posted April 21, 2011.   
3                                   NaN      Posted June 24, 2014.   
4                                   NaN  Posted September 8, 2013.   
...                                 ...                        ...   
59300                               NaN            Posted July 10.   
59301                               NaN             Posted July 8.   
59302  1 person found this review funny             Posted July 3.   
59303                               NaN            Posted July 20.   
59304                               NaN             Posted July 2.   

      last_edited   item_id                                          helpful  \
0             NaN    1250.0                                   No ratings yet   
1             NaN   22200.0                                   No ratings yet   
2             NaN   43110.0                                   No ratings yet   
3             NaN  251610.0  15 of 20 people (75%) found this review helpful   
4             NaN  227300.0     0 of 1 people (0%) found this review helpful   
...           ...       ...                                              ...   
59300         NaN      70.0                                   No ratings yet   
59301         NaN  362890.0                                   No ratings yet   
59302         NaN  273110.0    1 of 2 people (50%) found this review helpful   
59303         NaN     730.0                                   No ratings yet   
59304         NaN     440.0                                   No ratings yet   

      recommend  sentiment_analysis review_year  
0          True                   2        2011  
1          True                   2        2011  
2          True                   2        2011  
3          True                   2        2014  
4          True                   2        2013  
...         ...                 ...         ...  
59300      True                   2        2016  
59301      True                   2        2016  
59302      True                   2        2016  
59303      True                   2        2016  
59304      True                   2        2016  

[59305 rows x 10 columns]

**Step 2:**

df_steam_games has a column called `release_date`. We need to extract the year of this release date and place it in a new column called `release_year`. We ar going to do something similar as in the step 1.

In [21]:
df_steam_games

publisher                                             genres  \
0             Kotoshiro  ['Action', 'Casual', 'Indie', 'Simulation', 'S...   
1      Making Fun, Inc.       ['Free to Play', 'Indie', 'RPG', 'Strategy']   
2          Poolians.com  ['Casual', 'Free to Play', 'Indie', 'Simulatio...   
3                  彼岸领域                  ['Action', 'Adventure', 'Casual']   
4                   NaN                                                NaN   
...                 ...                                                ...   
32130   Ghost_RUS Games      ['Casual', 'Indie', 'Simulation', 'Strategy']   
32131            Sacada                    ['Casual', 'Indie', 'Strategy']   
32132      Laush Studio                  ['Indie', 'Racing', 'Simulation']   
32133          SIXNAILS                                ['Casual', 'Indie']   
32134               NaN                                                NaN   

                       app_name                     title  \
0           Lost Summoner Kitty       Lost Summoner Kitty   
1                     Ironbound                 Ironbound   
2       Real Pool 3D - Poolians   Real Pool 3D - Poolians   
3                       弹炸人2222                   弹炸人2222   
4                 Log Challenge                       NaN   
...                         ...                       ...   
32130            Colony On Mars            Colony On Mars   
32131  LOGistICAL: South Africa  LOGistICAL: South Africa   
32132             Russian Roads             Russian Roads   
32133       EXIT 2 - Directions       EXIT 2 - Directions   
32134               Maze Run VR                       NaN   

                                                     url release_date  \
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
1      http://store.steampowered.com/app/643980/Ironb...   2018-01-04   
2      http://store.steampowered.com/app/670290/Real_...   2017-07-24   
3         http://store.steampowered.com/app/767400/2222/   2017-12-07   
4      http://store.steampowered.com/app/773570/Log_C...          NaN   
...                                                  ...          ...   
32130  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
32131  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
32132  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
32133  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
32134  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                    tags  \
0      ['Strategy', 'Action', 'Indie', 'Casual', 'Sim...   
1      ['Free to Play', 'Strategy', 'Indie', 'RPG', '...   
2      ['Free to Play', 'Simulation', 'Sports', 'Casu...   
3                      ['Action', 'Adventure', 'Casual']   
4                ['Action', 'Indie', 'Casual', 'Sports']   
...                                                  ...   
32130      ['Strategy', 'Indie', 'Casual', 'Simulation']   
32131                    ['Strategy', 'Indie', 'Casual']   
32132                  ['Indie', 'Simulation', 'Racing']   
32133  ['Indie', 'Casual', 'Puzzle', 'Singleplayer', ...   
32134  ['Early Access', 'Adventure', 'Indie', 'Action...   

                                             reviews_url  \
0      http://steamcommunity.com/app/761140/reviews/?...   
1      http://steamcommunity.com/app/643980/reviews/?...   
2      http://steamcommunity.com/app/670290/reviews/?...   
3      http://steamcommunity.com/app/767400/reviews/?...   
4      http://steamcommunity.com/app/773570/reviews/?...   
...                                                  ...   
32130  http://steamcommunity.com/app/773640/reviews/?...   
32131  http://steamcommunity.com/app/733530/reviews/?...   
32132  http://steamcommunity.com/app/610660/reviews/?...   
32133  http://steamcommunity.com/app/658870/reviews/?...   
32134  http://steamcommunity.com/app/681550/reviews/?...   

                         

As in the step 1, it is a good practice to check the NaN in the column 'release_date'.

In [22]:
mask = df_steam_games['release_date'].isna()
df_steam_games[mask]

publisher                                            genres  \
4                   NaN                                               NaN   
11                  NaN                                               NaN   
19                  NaN                                               NaN   
20                  NaN                                               NaN   
22                  NaN                                               NaN   
...                 ...                                               ...   
32077               NaN                                               NaN   
32084               NaN                     ['Action', 'Casual', 'Indie']   
32085               NaN                                               NaN   
32088  rechargecomplete  ['Action', 'Adventure', 'Indie', 'Early Access']   
32134               NaN                                               NaN   

                   app_name              title  \
4             Log Challenge                NaN   
11     Icarus Six Sixty Six                NaN   
19            After Life VR                NaN   
20               Kitty Hawk                NaN   
22               Mortars VR                NaN   
...                     ...                ...   
32077           SpaceWalker                NaN   
32084              BOMBFEST           BOMBFEST   
32085            LIV Client                NaN   
32088     RECHARGE COMPLETE  RECHARGE COMPLETE   
32134           Maze Run VR                NaN   

                                                     url release_date  \
4      http://store.steampowered.com/app/773570/Log_C...          NaN   
11     http://store.steampowered.com/app/724910/Icaru...          NaN   
19     http://store.steampowered.com/app/772590/After...          NaN   
20     http://store.steampowered.com/app/640250/Kitty...          NaN   
22     http://store.steampowered.com/app/711440/Morta...          NaN   
...                                                  ...          ...   
32077  http://store.steampowered.com/app/705860/Space...          NaN   
32084  http://store.steampowered.com/app/729640/BOMBF...          NaN   
32085  http://store.steampowered.com/app/755540/LIV_C...          NaN   
32088  http://store.steampowered.com/app/708070/RECHA...          NaN   
32134  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                    tags  \
4                ['Action', 'Indie', 'Casual', 'Sports']   
11                                            ['Casual']   
19                       ['Early Access', 'Indie', 'VR']   
20     ['Early Access', 'Action', 'Adventure', 'Indie...   
22     ['Early Access', 'Strategy', 'Action', 'Indie'...   
...                                                  ...   
32077                         ['Early Access', 'Casual']   
32084                      ['Casual', 'Indie', 'Action']   
32085  ['Video Production', 'Utilities', 'Web Publish...   
32088   ['Early Access', 'Action', 'Adventure', 'Indie']   
32134  ['Early Access', 'Adventure', 'Indie', 'Action...   

                                             reviews_url  \
4      http://steamcommunity.com/app/773570/reviews/?...   
11     http://steamcommunity.com/app/724910/reviews/?...   
19     http://steamcommunity.com/app/772590/reviews/?...   
20     http://steamcommunity.com/app/640250/reviews/?...   
22     http://steamcommunity.com/app/711440/reviews/?...   
...                                                  ...   
32077  http://steamcommunity.com/app/705860/reviews/?...   
32084  http://steamcommunity.com/app/729640/reviews/?...   
32085  http://steamcommunity.com/app/755540/reviews/?...   
32088  http://steamcommunity.com/app/708070/reviews/?...   
32134  http://steamcommunity.com/app/681550/reviews/?...   

                                                   specs price  early_access  \
4      ['Single-player', 'Full controller support', '...  2.99         False   
11     ['

In this case, we can see that though the 'release_date' is in NaN, the entries have other information that can be useful.

Now, let's review the format of the dates in the column 'release_date', to check how can we extract the year.

In [23]:
df_steam_games['release_date'][:60]

0     2018-01-04
1     2018-01-04
2     2017-07-24
3     2017-12-07
4            NaN
5     2018-01-04
6     2018-01-04
7     2018-01-04
8     2018-01-04
9     2018-01-04
10        Soon..
11           NaN
12    2018-01-04
13    2018-01-04
14    2018-01-03
15    2018-01-03
16    2018-01-03
17    2018-01-03
18    2018-01-03
19           NaN
20           NaN
21    2018-01-03
22           NaN
23    2017-12-22
24    2017-12-23
25    2018-01-03
26    2018-01-03
27    1997-06-30
28    1998-11-08
29    2016-11-25
30           NaN
31    2018-01-01
32    2018-01-01
33    2018-01-01
34    2018-01-01
35    2017-12-30
36           NaN
37    2018-01-01
38    2018-01-01
39    2006-07-06
40    2006-07-06
41    2006-07-11
42           NaN
43          2017
44     Beta测试已开启
45           NaN
46    2017-12-29
47    2018-03-30
48    2017-12-29
49    2017-12-29
50    2017-12-29
51           NaN
52    2017-12-30
53           NaN
54    2017-12-29
55    2005-08-09
56    2006-09-29
57    2006-11-20
58    2006-11-

In the sample of the first 60 rows we can see that the normal format is YYYY-MM-DD. However, there are some NaN and strings that need to be considered. Also, there are entries with only the year. So we need to consider also this.

In [24]:
# We create a function to extract the year considering the data that we have in the sample above.
def extract_year(date_str):
    if pd.notna(date_str):  #If the value is not NaN
        year_match = re.search(r'\b\d{4}\b', date_str) # We loo for the year following the format '\b\d{4}\b' using the function re.search()
        if year_match:
            return year_match.group()
    return np.NaN

# we apply the function to the column release_date
release_year = df_steam_games['release_date'].apply(extract_year)

# Mostrar el DataFrame resultante
release_year

0        2018
1        2018
2        2017
3        2017
4         NaN
         ... 
32130    2018
32131    2018
32132    2018
32133    2017
32134     NaN
Name: release_date, Length: 32135, dtype: object

Now we concatenate the column release_year with the dataframe steam_games:

In [25]:
df_steam_games_WY = pd.concat([df_steam_games, pd.DataFrame({'release_year': release_year})], axis=1)
df_steam_games_WY

publisher                                             genres  \
0             Kotoshiro  ['Action', 'Casual', 'Indie', 'Simulation', 'S...   
1      Making Fun, Inc.       ['Free to Play', 'Indie', 'RPG', 'Strategy']   
2          Poolians.com  ['Casual', 'Free to Play', 'Indie', 'Simulatio...   
3                  彼岸领域                  ['Action', 'Adventure', 'Casual']   
4                   NaN                                                NaN   
...                 ...                                                ...   
32130   Ghost_RUS Games      ['Casual', 'Indie', 'Simulation', 'Strategy']   
32131            Sacada                    ['Casual', 'Indie', 'Strategy']   
32132      Laush Studio                  ['Indie', 'Racing', 'Simulation']   
32133          SIXNAILS                                ['Casual', 'Indie']   
32134               NaN                                                NaN   

                       app_name                     title  \
0           Lost Summoner Kitty       Lost Summoner Kitty   
1                     Ironbound                 Ironbound   
2       Real Pool 3D - Poolians   Real Pool 3D - Poolians   
3                       弹炸人2222                   弹炸人2222   
4                 Log Challenge                       NaN   
...                         ...                       ...   
32130            Colony On Mars            Colony On Mars   
32131  LOGistICAL: South Africa  LOGistICAL: South Africa   
32132             Russian Roads             Russian Roads   
32133       EXIT 2 - Directions       EXIT 2 - Directions   
32134               Maze Run VR                       NaN   

                                                     url release_date  \
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
1      http://store.steampowered.com/app/643980/Ironb...   2018-01-04   
2      http://store.steampowered.com/app/670290/Real_...   2017-07-24   
3         http://store.steampowered.com/app/767400/2222/   2017-12-07   
4      http://store.steampowered.com/app/773570/Log_C...          NaN   
...                                                  ...          ...   
32130  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
32131  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
32132  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
32133  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
32134  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                    tags  \
0      ['Strategy', 'Action', 'Indie', 'Casual', 'Sim...   
1      ['Free to Play', 'Strategy', 'Indie', 'RPG', '...   
2      ['Free to Play', 'Simulation', 'Sports', 'Casu...   
3                      ['Action', 'Adventure', 'Casual']   
4                ['Action', 'Indie', 'Casual', 'Sports']   
...                                                  ...   
32130      ['Strategy', 'Indie', 'Casual', 'Simulation']   
32131                    ['Strategy', 'Indie', 'Casual']   
32132                  ['Indie', 'Simulation', 'Racing']   
32133  ['Indie', 'Casual', 'Puzzle', 'Singleplayer', ...   
32134  ['Early Access', 'Adventure', 'Indie', 'Action...   

                                             reviews_url  \
0      http://steamcommunity.com/app/761140/reviews/?...   
1      http://steamcommunity.com/app/643980/reviews/?...   
2      http://steamcommunity.com/app/670290/reviews/?...   
3      http://steamcommunity.com/app/767400/reviews/?...   
4      http://steamcommunity.com/app/773570/reviews/?...   
...                                                  ...   
32130  http://steamcommunity.com/app/773640/reviews/?...   
32131  http://steamcommunity.com/app/733530/reviews/?...   
32132  http://steamcommunity.com/app/610660/reviews/?...   
32133  http://steamcommunity.com/app/658870/reviews/?...   
32134  http://steamcommunity.com/app/681550/reviews/?...   

                         

**Step 3:**

Having a dataframe with the item IDs for that specific `genre`, we can merge it with a dataframe that has the sum of 'playtime_forever' grouped by `item_id`. This information is found in the dataframe df_users_items. Now we have a new dataframe for a specific genre with the sum of hours played.

In [26]:
df_users_items

user_id  items_count           steam_id  \
0        76561197970982479          277  76561197970982479   
1        76561197970982479          277  76561197970982479   
2        76561197970982479          277  76561197970982479   
3        76561197970982479          277  76561197970982479   
4        76561197970982479          277  76561197970982479   
...                    ...          ...                ...   
5170010  76561198329548331            7  76561198329548331   
5170011  76561198329548331            7  76561198329548331   
5170012  76561198329548331            7  76561198329548331   
5170013  76561198329548331            7  76561198329548331   
5170014  edward_tremethick            0  76561198331598578   

                                                  user_url   item_id  \
0        http://steamcommunity.com/profiles/76561197970...      10.0   
1        http://steamcommunity.com/profiles/76561197970...      20.0   
2        http://steamcommunity.com/profiles/76561197970...      30.0   
3        http://steamcommunity.com/profiles/76561197970...      40.0   
4        http://steamcommunity.com/profiles/76561197970...      50.0   
...                                                    ...       ...   
5170010  http://steamcommunity.com/profiles/76561198329...  373330.0   
5170011  http://steamcommunity.com/profiles/76561198329...  388490.0   
5170012  http://steamcommunity.com/profiles/76561198329...  521570.0   
5170013  http://steamcommunity.com/profiles/76561198329...  519140.0   
5170014     http://steamcommunity.com/id/edward_tremethick       NaN   

                             item_name  playtime_forever  playtime_2weeks  
0                       Counter-Strike               6.0              0.0  
1                Team Fortress Classic               0.0              0.0  
2                        Day of Defeat               7.0              0.0  
3                   Deathmatch Classic               0.0              0.0  
4            Half-Life: Opposing Force               0.0              0.0  
...                                ...               ...              ...  
5170010                    All Is Dust               0.0              0.0  
5170011  One Way To Die: Steam Edition               3.0              3.0  
5170012          You Have 10 Seconds 2               4.0              4.0  
5170013                     Minds Eyes               3.0              3.0  
5170014                            NaN               NaN              NaN  

[5170015 rows x 8 columns]

Let's group by item_id summing the playtime_forever column. To so this, we can use the function "groupby()" of Pandas:

In [27]:
# Group by "item_id" and sum the values of "playtime_forever"
grouped_df_users_items = df_users_items.groupby(['item_id', 'user_id'])['playtime_forever'].sum().reset_index()

grouped_df_users_items.rename(columns={'playtime_forever': 'sum_playtime_forever'}, inplace=True)
grouped_df_users_items

item_id               user_id  sum_playtime_forever
0            10.0            -GM-Dragon                  14.0
1            10.0            -KillZone-                 987.0
2            10.0                 -Mad-                   8.0
3            10.0               -SEVEN-                 617.0
4            10.0  -SatansLittleHelper-                 511.0
...           ...                   ...                   ...
5094077  529670.0                chidvd                   0.0
5094078  529670.0        fearlesskeeper                   0.0
5094079  529820.0           robinwalker                   0.0
5094080  530720.0            CSMisBeast                   0.0
5094081  530720.0        mayshowganmore                   0.0

[5094082 rows x 3 columns]

We passed from 5,170,015 entries to 5,094,082 registers.

Now we can do a merge between the dataframes grouped_df_users_items and df_user_reviews_WY using the item IDs.


In [28]:
merged_uitems_ureviews = pd.merge(grouped_df_users_items, df_user_reviews_WY, on=['item_id', 'user_id'], how='inner')
merged_uitems_ureviews

item_id            user_id  sum_playtime_forever  \
0          10.0           71251241                 566.0   
1          10.0  76561198015886143              144786.0   
2          10.0  76561198040188061                3895.0   
3          10.0  76561198072207162                  54.0   
4          10.0  76561198076217855                3848.0   
...         ...                ...                   ...   
44885  521430.0  76561198075141715                   5.0   
44886  521570.0  76561198071122396                  25.0   
44887  521570.0        TfhuAWGscvg                  54.0   
44888  521990.0           mikeyg74                  16.0   
44889  527340.0           Limo1300                 147.0   

                                                user_url  \
0                  http://steamcommunity.com/id/71251241   
1      http://steamcommunity.com/profiles/76561198015...   
2      http://steamcommunity.com/profiles/76561198040...   
3      http://steamcommunity.com/profiles/76561198072...   
4      http://steamcommunity.com/profiles/76561198076...   
...                                                  ...   
44885  http://steamcommunity.com/profiles/76561198075...   
44886  http://steamcommunity.com/profiles/76561198071...   
44887           http://steamcommunity.com/id/TfhuAWGscvg   
44888              http://steamcommunity.com/id/mikeyg74   
44889              http://steamcommunity.com/id/Limo1300   

                                  funny                    posted last_edited  \
0                                   NaN  Posted November 7, 2015.         NaN   
1                                   NaN     Posted June 14, 2015.         NaN   
2                                   NaN      Posted May 18, 2011.         NaN   
3                                   NaN  Posted January 23, 2014.         NaN   
4                                   NaN      Posted May 18, 2014.         NaN   
...                                 ...                       ...         ...   
44885  2 people found this review funny         Posted August 25.         NaN   
44886                               NaN         Posted August 29.         NaN   
44887  1 person found this review funny         Posted August 30.         NaN   
44888                               NaN      Posted September 17.         NaN   
44889  4 people found this review funny      Posted September 15.         NaN   

                                              helpful recommend  \
0                                      No ratings yet      True   
1      1 of 1 people (100%) found this review helpful      True   
2        0 of 2 people (0%) found this review helpful      True   
3                                      No ratings yet      True   
4                                      No ratings yet      True   
...                                               ...       ...   
44885   3 of 5 people (60%) found this review helpful      True   
44886                                  No ratings yet      True   
44887                                  No ratings yet      True   
44888                                  No ratings yet      True   
44889   2 of 4 people (50%) found this review helpful      True   

       sentiment_analysis review_year  
0                       2        2015  
1                       1        2015  
2                       2        2011  
3                       2        2014  
4                       2        2014  
...                   ...         ...  
44885                   2        2016  
44886                   0        2016  
44887                   0        2016  
44888                   2        2016  
44889                   0        2016  

[44890 rows x 11 columns]

Now we have a dataframe with the "item_id", the sum of "playtime_forever" and segregated by "user_id". Also, we have the year when that user played that "item_id" and made a review.

Now we need to merge the last merged dataframe with the df_steam_games_WY in order to have all the information together and apply the functions:

In [29]:
df_steam_games_WY.rename(columns={'id': 'item_id'}, inplace=True)
df_steam_games_WY

publisher                                             genres  \
0             Kotoshiro  ['Action', 'Casual', 'Indie', 'Simulation', 'S...   
1      Making Fun, Inc.       ['Free to Play', 'Indie', 'RPG', 'Strategy']   
2          Poolians.com  ['Casual', 'Free to Play', 'Indie', 'Simulatio...   
3                  彼岸领域                  ['Action', 'Adventure', 'Casual']   
4                   NaN                                                NaN   
...                 ...                                                ...   
32130   Ghost_RUS Games      ['Casual', 'Indie', 'Simulation', 'Strategy']   
32131            Sacada                    ['Casual', 'Indie', 'Strategy']   
32132      Laush Studio                  ['Indie', 'Racing', 'Simulation']   
32133          SIXNAILS                                ['Casual', 'Indie']   
32134               NaN                                                NaN   

                       app_name                     title  \
0           Lost Summoner Kitty       Lost Summoner Kitty   
1                     Ironbound                 Ironbound   
2       Real Pool 3D - Poolians   Real Pool 3D - Poolians   
3                       弹炸人2222                   弹炸人2222   
4                 Log Challenge                       NaN   
...                         ...                       ...   
32130            Colony On Mars            Colony On Mars   
32131  LOGistICAL: South Africa  LOGistICAL: South Africa   
32132             Russian Roads             Russian Roads   
32133       EXIT 2 - Directions       EXIT 2 - Directions   
32134               Maze Run VR                       NaN   

                                                     url release_date  \
0      http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
1      http://store.steampowered.com/app/643980/Ironb...   2018-01-04   
2      http://store.steampowered.com/app/670290/Real_...   2017-07-24   
3         http://store.steampowered.com/app/767400/2222/   2017-12-07   
4      http://store.steampowered.com/app/773570/Log_C...          NaN   
...                                                  ...          ...   
32130  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
32131  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
32132  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
32133  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
32134  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                    tags  \
0      ['Strategy', 'Action', 'Indie', 'Casual', 'Sim...   
1      ['Free to Play', 'Strategy', 'Indie', 'RPG', '...   
2      ['Free to Play', 'Simulation', 'Sports', 'Casu...   
3                      ['Action', 'Adventure', 'Casual']   
4                ['Action', 'Indie', 'Casual', 'Sports']   
...                                                  ...   
32130      ['Strategy', 'Indie', 'Casual', 'Simulation']   
32131                    ['Strategy', 'Indie', 'Casual']   
32132                  ['Indie', 'Simulation', 'Racing']   
32133  ['Indie', 'Casual', 'Puzzle', 'Singleplayer', ...   
32134  ['Early Access', 'Adventure', 'Indie', 'Action...   

                                             reviews_url  \
0      http://steamcommunity.com/app/761140/reviews/?...   
1      http://steamcommunity.com/app/643980/reviews/?...   
2      http://steamcommunity.com/app/670290/reviews/?...   
3      http://steamcommunity.com/app/767400/reviews/?...   
4      http://steamcommunity.com/app/773570/reviews/?...   
...                                                  ...   
32130  http://steamcommunity.com/app/773640/reviews/?...   
32131  http://steamcommunity.com/app/733530/reviews/?...   
32132  http://steamcommunity.com/app/610660/reviews/?...   
32133  http://steamcommunity.com/app/658870/reviews/?...   
32134  http://steamcommunity.com/app/681550/reviews/?...   

                         

In [30]:
#Merging all the dataframes by item_id.
merged_all_by_sumPTForever = pd.merge(merged_uitems_ureviews, df_steam_games_WY, on='item_id', how='inner')
merged_all_by_sumPTForever

item_id            user_id  sum_playtime_forever  \
0          10.0           71251241                 566.0   
1          10.0  76561198015886143              144786.0   
2          10.0  76561198040188061                3895.0   
3          10.0  76561198072207162                  54.0   
4          10.0  76561198076217855                3848.0   
...         ...                ...                   ...   
40743  521430.0  76561198075141715                   5.0   
40744  521570.0  76561198071122396                  25.0   
40745  521570.0        TfhuAWGscvg                  54.0   
40746  521990.0           mikeyg74                  16.0   
40747  527340.0           Limo1300                 147.0   

                                                user_url  \
0                  http://steamcommunity.com/id/71251241   
1      http://steamcommunity.com/profiles/76561198015...   
2      http://steamcommunity.com/profiles/76561198040...   
3      http://steamcommunity.com/profiles/76561198072...   
4      http://steamcommunity.com/profiles/76561198076...   
...                                                  ...   
40743  http://steamcommunity.com/profiles/76561198075...   
40744  http://steamcommunity.com/profiles/76561198071...   
40745           http://steamcommunity.com/id/TfhuAWGscvg   
40746              http://steamcommunity.com/id/mikeyg74   
40747              http://steamcommunity.com/id/Limo1300   

                                  funny                    posted last_edited  \
0                                   NaN  Posted November 7, 2015.         NaN   
1                                   NaN     Posted June 14, 2015.         NaN   
2                                   NaN      Posted May 18, 2011.         NaN   
3                                   NaN  Posted January 23, 2014.         NaN   
4                                   NaN      Posted May 18, 2014.         NaN   
...                                 ...                       ...         ...   
40743  2 people found this review funny         Posted August 25.         NaN   
40744                               NaN         Posted August 29.         NaN   
40745  1 person found this review funny         Posted August 30.         NaN   
40746                               NaN      Posted September 17.         NaN   
40747  4 people found this review funny      Posted September 15.         NaN   

                                              helpful recommend  \
0                                      No ratings yet      True   
1      1 of 1 people (100%) found this review helpful      True   
2        0 of 2 people (0%) found this review helpful      True   
3                                      No ratings yet      True   
4                                      No ratings yet      True   
...                                               ...       ...   
40743   3 of 5 people (60%) found this review helpful      True   
40744                                  No ratings yet      True   
40745                                  No ratings yet      True   
40746                                  No ratings yet      True   
40747   2 of 4 people (50%) found this review helpful      True   

       sentiment_analysis  ...                  title  \
0                       2  ...         Counter-Strike   
1                       1  ...         Counter-Strike   
2                       2  ...         Counter-Strike   
3                       2  ...         Counter-Strike   
4                       2  ...         Counter-Strike   
...                   ...  ...                    ...   
40743                   2  ...           Super Switch   
40744                   0  ...  You Have 10 Seconds 2   
40745                   0  ...  You Have 10 Seconds 2   
40746                   2  ...         Galactic Storm   
40747                   0  ...          What The Box?   

                                                     url release_date  \
0      http://store.steampower

In [31]:
merged_all_by_sumPTForever.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 40748 entries, 0 to 40747
Data columns (total 24 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   item_id               40748 non-null  float64
 1   user_id               40748 non-null  object 
 2   sum_playtime_forever  40748 non-null  float64
 3   user_url              40748 non-null  object 
 4   funny                 5874 non-null   object 
 5   posted                40748 non-null  object 
 6   last_edited           4225 non-null   object 
 7   helpful               40748 non-null  object 
 8   recommend             40748 non-null  object 
 9   sentiment_analysis    40748 non-null  int64  
 10  review_year           40748 non-null  object 
 11  publisher             38720 non-null  object 
 12  genres                39071 non-null  object 
 13  app_name              40748 non-null  object 
 14  title                 39205 non-null  object 
 15  url                

Now we need to keep minimum columns that are requiered by the functions in order to save resources.

Analyzing the functions, the only columns that we are going to need are the following:

- item_id: It is important for the ML model where the user needs to enter an item_id in order to receive 5 games that are similar to the one selected. Also, it is needed for the functions 3 and 4.
- title: It is important for the ML model where the user needs to enter an item_id in order to receive 5 games that are similar to the one selected. Also, it is needed for the functions 3 and 4.
- release_year: We need these data for the function 5.
- genres: This is a good way to find similar items in the ML model.
- user_id: It is important for the ML model where the user needs to enter a user_id in order to receive 5 games that are similar to the one selected.
- sum_playtime_forever: This data is important because if the game is played for more time by any user, is. because he/she has more engagement with it and may like similar ones.
- recommend: This data is needed for the functions 3 and 4.
- sentiment_analysis: This data is needed for the functions 3 and 4.
- reviews_year: This data is needed for the functions 3 and 4.


In [32]:
selected_columns = ['item_id','title','release_year','genres','user_id','sum_playtime_forever','recommend','sentiment_analysis','review_year']
df_F345 = merged_all_by_sumPTForever[selected_columns]
df_F345

item_id                  title release_year  \
0          10.0         Counter-Strike         2000   
1          10.0         Counter-Strike         2000   
2          10.0         Counter-Strike         2000   
3          10.0         Counter-Strike         2000   
4          10.0         Counter-Strike         2000   
...         ...                    ...          ...   
40743  521430.0           Super Switch         2016   
40744  521570.0  You Have 10 Seconds 2         2016   
40745  521570.0  You Have 10 Seconds 2         2016   
40746  521990.0         Galactic Storm         2016   
40747  527340.0          What The Box?         2016   

                                    genres            user_id  \
0                               ['Action']           71251241   
1                               ['Action']  76561198015886143   
2                               ['Action']  76561198040188061   
3                               ['Action']  76561198072207162   
4                               ['Action']  76561198076217855   
...                                    ...                ...   
40743                            ['Indie']  76561198075141715   
40744  ['Casual', 'Free to Play', 'Indie']  76561198071122396   
40745  ['Casual', 'Free to Play', 'Indie']        TfhuAWGscvg   
40746                  ['Action', 'Indie']           mikeyg74   
40747                  ['Action', 'Indie']           Limo1300   

       sum_playtime_forever recommend  sentiment_analysis review_year  
0                     566.0      True                   2        2015  
1                  144786.0      True                   1        2015  
2                    3895.0      True                   2        2011  
3                      54.0      True                   2        2014  
4                    3848.0      True                   2        2014  
...                     ...       ...                 ...         ...  
40743                   5.0      True                   2        2016  
40744                  25.0      True                   0        2016  
40745                  54.0      True                   0        2016  
40746                  16.0      True                   2        2016  
40747                 147.0      True                   0        2016  

[40748 rows x 9 columns]

Let's change the type of some columns in order to avoid problems with the functions:

In [33]:
df_F345.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 40748 entries, 0 to 40747
Data columns (total 9 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   item_id               40748 non-null  float64
 1   title                 39205 non-null  object 
 2   release_year          39149 non-null  object 
 3   genres                39071 non-null  object 
 4   user_id               40748 non-null  object 
 5   sum_playtime_forever  40748 non-null  float64
 6   recommend             40748 non-null  object 
 7   sentiment_analysis    40748 non-null  int64  
 8   review_year           40748 non-null  object 
dtypes: float64(2), int64(1), object(6)
memory usage: 3.1+ MB


In [35]:
df_F345['title'] = df_F345['title'].astype('string')
df_F345.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 40748 entries, 0 to 40747
Data columns (total 9 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   item_id               40748 non-null  float64
 1   title                 39205 non-null  string 
 2   release_year          39149 non-null  object 
 3   genres                39071 non-null  object 
 4   user_id               40748 non-null  object 
 5   sum_playtime_forever  40748 non-null  float64
 6   recommend             40748 non-null  object 
 7   sentiment_analysis    40748 non-null  int64  
 8   review_year           40748 non-null  object 
dtypes: float64(2), int64(1), object(5), string(1)
memory usage: 3.1+ MB


<ipython-input-35-4b34578cf981>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_F345['title'] = df_F345['title'].astype('string')


In [36]:
df_F345['review_year'] = df_F345['review_year'].astype('int64')
df_F345.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 40748 entries, 0 to 40747
Data columns (total 9 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   item_id               40748 non-null  float64
 1   title                 39205 non-null  string 
 2   release_year          39149 non-null  object 
 3   genres                39071 non-null  object 
 4   user_id               40748 non-null  object 
 5   sum_playtime_forever  40748 non-null  float64
 6   recommend             40748 non-null  object 
 7   sentiment_analysis    40748 non-null  int64  
 8   review_year           40748 non-null  int64  
dtypes: float64(2), int64(2), object(4), string(1)
memory usage: 3.1+ MB


<ipython-input-36-637dbe7b33f2>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_F345['review_year'] = df_F345['review_year'].astype('int64')


We eliminate NaN values from the "release_year" column.

In [37]:
df_F345.dropna(subset='release_year',inplace=True)
df_F345

<ipython-input-37-6de88864d258>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_F345.dropna(subset='release_year',inplace=True)


item_id                  title release_year  \
0          10.0         Counter-Strike         2000   
1          10.0         Counter-Strike         2000   
2          10.0         Counter-Strike         2000   
3          10.0         Counter-Strike         2000   
4          10.0         Counter-Strike         2000   
...         ...                    ...          ...   
40743  521430.0           Super Switch         2016   
40744  521570.0  You Have 10 Seconds 2         2016   
40745  521570.0  You Have 10 Seconds 2         2016   
40746  521990.0         Galactic Storm         2016   
40747  527340.0          What The Box?         2016   

                                    genres            user_id  \
0                               ['Action']           71251241   
1                               ['Action']  76561198015886143   
2                               ['Action']  76561198040188061   
3                               ['Action']  76561198072207162   
4                               ['Action']  76561198076217855   
...                                    ...                ...   
40743                            ['Indie']  76561198075141715   
40744  ['Casual', 'Free to Play', 'Indie']  76561198071122396   
40745  ['Casual', 'Free to Play', 'Indie']        TfhuAWGscvg   
40746                  ['Action', 'Indie']           mikeyg74   
40747                  ['Action', 'Indie']           Limo1300   

       sum_playtime_forever recommend  sentiment_analysis  review_year  
0                     566.0      True                   2         2015  
1                  144786.0      True                   1         2015  
2                    3895.0      True                   2         2011  
3                      54.0      True                   2         2014  
4                    3848.0      True                   2         2014  
...                     ...       ...                 ...          ...  
40743                   5.0      True                   2         2016  
40744                  25.0      True                   0         2016  
40745                  54.0      True                   0         2016  
40746                  16.0      True                   2         2016  
40747                 147.0      True                   0         2016  

[39149 rows x 9 columns]

In [38]:
df_F345['release_year'] = df_F345['release_year'].astype('int64')
df_F345.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 39149 entries, 0 to 40747
Data columns (total 9 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   item_id               39149 non-null  float64
 1   title                 39149 non-null  string 
 2   release_year          39149 non-null  int64  
 3   genres                39015 non-null  object 
 4   user_id               39149 non-null  object 
 5   sum_playtime_forever  39149 non-null  float64
 6   recommend             39149 non-null  object 
 7   sentiment_analysis    39149 non-null  int64  
 8   review_year           39149 non-null  int64  
dtypes: float64(2), int64(3), object(3), string(1)
memory usage: 3.0+ MB


<ipython-input-38-7eec4f5c102b>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_F345['release_year'] = df_F345['release_year'].astype('int64')


We are going to save the entire dataset in a CSV file to have a backup copy ready to be consumed in the next processes.

In [39]:
df_F345.to_csv('F345ML.csv', index=False)

Let's save it also in a parquet file due to it saves resources.

In [40]:
df_F345.to_parquet('F345.parquet', index=False)